Update parse.py for better code and testing

diivm opened the pull request:
- move utitlity functions to `utils.py`
- remove terminal color print
- simplify `print_node`
- change `node_in_this_file` arg name
- remove `filter` and `lines` from functions: unused variables
- change `generate_parsed_info` return type to dict() from None
- changed `generate_parsed_info` operations from list to dict
- add `create_index`
- add `get_compilation_database`
- add `get_compile_commands`
- use real paths
- use`utils.py` functions
- use named arguments
- fix paths after moving to `scripts/`
- use named args

diivm committed:
[utils.py] add utils

diivm committed:
[parse.py] clean and simplify

- move utitlity functions to `utils.py`
- remove terminal color print
- simplify `print_node`
- change `node_in_this_file` arg name
- remove `filter` and `lines` from functions: unused variables
- change `generate_parsed_info` return type to dict() from None
- changed `generate_parsed_info` operations from list to dict
- add `create_index`
- add `get_compilation_database`
- add `get_compile_commands`
- use real paths
- use`utils.py` functions
- use named arguments
- fix paths after moving to `scripts/`
- use named args

divmadan commented:
![](https://github.trello.services/images/mini-trello-icon.png) [update parse.py](https://trello.com/c/NmOk4NnP/45-update-parsepy)

aPonza commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,89 @@
-
-
-def walk_and_print(cursor, filter, lines, this_filename, depth):
+def node_in_this_file(node, filename):
```
`bool` returning functions usually start with is (/has)

diivm replied:
Done.

aPonza commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,89 @@
-    parser.add_argument("files", nargs="+", help="The source files to search")
-    return parser.parse_args(args)
 
+def create_index():
```
seems unnecessary

diivm replied:
I can remove it if we don't want to test it. I modified the main function so that it calls either the file's(`parse.py`) functions or `utils.py`'s functions so that everything the main function does, can be tested.

aPonza replied:
Since this is only a call to an external function, you'd be testing their implementation, not yours. It wouldn't be useful to prevent your mistakes, only theirs

diivm replied:
Understood. Removed.

aPonza commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,89 @@
-        os.makedirs(dir)
 
-    return f"{dir}/{filename}.json"
+def get_compile_commands(compdb, filename):
```
Consistent names are better than short names (`compdb` -> `compile_db` for consistency with L58's `compile_db_path` or even `compilation_database` and `compilation_database_path` on L58.

diivm replied:
Done.

aPonza commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,89 @@
-            os.path.realpath(source), output_dir=f"json/{os.path.dirname(__file__)}"
-        )
-        dump_json(output_filepath, parsed_list)
+        compile_db_path = utils.get_dirname(path="../compile_commands.json")
```
Will relative paths ever be an issue?

diivm replied:
Good point. They are while running the code from different directories.
Something like b036420 should work.

aPonza replied:
Eh, now the code only works on your machine though, since e.g. CI runs somewhere that doesn't have `/home/divyanshu/Projects/active` prepended to the `/pcl`-bit

diivm replied:
A suggestion would be appreciated :slightly_smiling_face:

kunaltyagi replied:
Use `os.path` or something similar in Python. I'd recommend allowing the user to provide the script, and having a sane default

diivm replied:
I let the user specify the path to `compile_commands.json` and the directory where they want to store the output JSON(s), the defaults for both being the parent directory of the `scripts`.


10b7c55 and 50a3a06

aPonza replied:
I googled "relative path python", [this is the first result](https://stackoverflow.com/questions/918154/relative-paths-in-python) and it seems good enough to me. Do notice some of the comments on portability, as CI runs on Windows too.

diivm replied:
I've used the same as the link :slightly_smiling_face:

aPonza commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,89 @@
-        dump_json(output_filepath, parsed_list)
+        compile_db_path = utils.get_dirname(path="../compile_commands.json")
+        compdb = get_compilation_database(compile_db_path=compile_db_path)
+        compile_commands = get_compile_commands(compdb=compdb, filename=source)
```
Why not a single function to get compilation flags? Any advantage in having three distinct?

diivm replied:
Initially, I thought there will be more uses of `compilation_database`. 
Merged into one.

diivm committed:
[parse.py] change bool function name

diivm committed:
[parse.py] consistent naming

diivm committed:
[parse.py] compilation_commands in one function and fixed path

diivm committed:
[parse.py] fixed path

diivm committed:
parse_arguments fix

diivm committed:
[utils.py] remove unused function

kunaltyagi commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,86 @@
-
-
-def walk_and_print(cursor, filter, lines, this_filename, depth):
+def is_node_in_this_file(node, filename):
```
What's node?

diivm replied:
`Node` of the AST. 
Change it to `is_cursor_in_this_file`?

diivm replied:
Changed argument name from `node` to  `cursor`

kunaltyagi replied:
Err.. what's cursor? :laughing: `print_ast` is a beautiful name (even without being compared to `walk_and_print`) and I'm sure you'll find the right name here too

kunaltyagi commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,86 @@
+    return node.location.file and node.location.file.name == filename
+
+
+def walk_and_print(cursor, this_filename, depth):
```
What are we walking?

diivm replied:
The AST :joy: 
Change to `traverse_and_print`?

diivm replied:
Changed to `print_ast`

kunaltyagi commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,86 @@
+import utils
+
+
+def print_node(cursor, depth):
```
docstrings are missing

diivm replied:
Noted. Will update.

diivm replied:
Updated in fa0a0cb

diivm committed:
[parse.py] remove `None` returns

diivm committed:
[parse.py] changes

- rename node argument to cursor
- remove print_node function
- rename walk_and_print to print_ast

diivm committed:
[parse.py] format using default args

diivm committed:
[parse.py] remove unnecessary create_index

diivm committed:
[parse.py] add docstrings

kunaltyagi commented on bindings/python/scripts/utils.py:
```diff
@@ -0,0 +1,39 @@
+
+    ensure_dir_exists(dir)
+
+    return get_realpath(f"{dir}/{filename}.json")
```
Use the path join operations in Pyhton to allow a Windows user to run this. Same in L21

diivm replied:
Done

kunaltyagi commented on bindings/python/scripts/utils.py:
```diff
@@ -0,0 +1,39 @@
+        json.dump(info, f, indent=2)
+
+
+def parse_arguments(args, script):
```
Any need of `script`?

diivm replied:
The `json -> pybind` script also uses `parse_arguments`, planning to add it's support.

kunaltyagi commented on bindings/python/scripts/utils.py:
```diff
@@ -0,0 +1,39 @@
+    if script == "parse":
+        parser = argparse.ArgumentParser(description="C++ libclang parser")
+        parser.add_argument("files", nargs="+", help="The source files to parse")
+        return parser.parse_args(args)
```
IIRC, you can also skip `args` and Python will automatically use sane defaults

diivm replied:
Done.

kunaltyagi commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,97 @@
-        with open(source) as input_file:
-            lines = input_file.readlines()
+        compilation_commands = get_compilation_commands(
+            compilation_database_path="/home/divyanshu/Projects/active/pcl/bindings/python",
```
Please get a dynamic path from user, and use a sane default:
* [temporary file](https://docs.python.org/3.6/library/tempfile.html)
* current directory

diivm replied:
https://github.com/divmadan/pcl/pull/10#discussion_r450897720
How about 10b7c55 and 50a3a06

kunaltyagi commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,97 @@
-        # extracting argument list from the command's generator object
-        compile_commands = list(compile_commands[0].arguments)[1:-1]
-        tu = index.parse(source, args=compile_commands)
+        tu = index.parse(path=source, args=compilation_commands)
```
You don't need to use `translation_unit` (and shouldn't) since the file might be a header (based on your design and usecase).

`file_ast` might be a good starting point for an alternative

diivm replied:
Header and impl files too can be compiled, right?
Also, this is only about the variable name, correct?

kunaltyagi replied:
`tu` has a very specific meaning and you're not using that context, so it's better to avoid confusion

diivm replied:
Good point. Changed to `source_ast`.

kunaltyagi commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,97 @@
-            os.path.realpath(source), output_dir=f"json/{os.path.dirname(__file__)}"
+        output_filepath = utils.get_json_output_path(
+            source=source,
+            output_dir="/home/divyanshu/Projects/active/pcl/bindings/python/json",
```
Same logic regarding paths as above

kunaltyagi reviewed (changes requested):
Haven't reviewed parse.py

diivm committed:
[utils.py] add compilation_database_path and json_output_path

diivm committed:
[parse.py] take user input for compilation_database_path and json_output_path

diivm committed:
[utils.py] changes

- join_path function
- add `\` delimiter for Windows
- use join_paths instead of '\'

diivm committed:
[parse.py] use join_path

diivm committed:
[parse.py] change `tu` to `source_ast`

diivm committed:
[utils.py] remove unnecessary `args` parameter

diivm committed:
[parse.py] update

kunaltyagi commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,143 @@
+    )
+    compilation_commands = compilation_database.getCompileCommands(filename=filename)
+    # extracting argument list from the command's generator object
+    return list(compilation_commands[0].arguments)[1:-1]
```
Comments on why are more important. In this case, why choose 0, 1 and -1

diivm replied:
Updated the function: c6b4ab3.

Same for the comments below.

diivm replied:
Updated

kunaltyagi commented on bindings/python/scripts/utils.py:
```diff
@@ -0,0 +1,57 @@
+
+
+def get_json_output_path(source, output_dir):
+    delimiter = "/" if "/" in source else "\\"
```
Python has an os delimiter (separator)

kunaltyagi commented on bindings/python/scripts/utils.py:
```diff
@@ -0,0 +1,57 @@
+
+def get_json_output_path(source, output_dir):
+    delimiter = "/" if "/" in source else "\\"
+    x_list = source.split(f"pcl{delimiter}", 1)[-1]
```
Python has an `os.path.split`

diivm committed:
[utils.py] update get_json_output_path

divmadan commented:
Do the functions in `utils.py` also require docstrings?

kunaltyagi commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,143 @@
 
+    Arguments:
+        - cursor: The cursor pointing to a node
+        - this_filename: The file's name to check if the node belongs to it
```
Please specify why this is needed (header files also get parsed)

diivm replied:
Done.

kunaltyagi commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,143 @@
+            print_ast(cursor=child, this_filename=this_filename, depth=depth + 1)
+
+
+def generate_parsed_info(cursor, this_filename, depth):
```
`cursor` and `this_filename` seem to go together. Maybe make a struct, tuple or something to make them go together.

This is also because the cursor needs to point in to the file requested (which you aren't checking at the beginning of the function)

diivm replied:
> Maybe make a struct, tuple or something to make them go together.

Done.

 > which you aren't checking at the beginning of the function

I tried, but the root (`TRANSLATION_UNIT`) has `cursor.location.file` as NoneType, so it leads to an empty json.

kunaltyagi commented on bindings/python/scripts/utils.py:
```diff
@@ -0,0 +1,75 @@
+        json.dump(info, f, indent=2)
+
+
+def parse_arguments(script):
```
Some functions definitely need docstrings, even in utils.py

diivm replied:
Done.

diivm committed:
[parse.py] refactor for better testing interface

kunaltyagi commented on bindings/python/scripts/parse.py:
```diff
@@ -115,28 +115,38 @@ def get_compilation_commands(compilation_database_path, filename):
-        source = utils.get_realpath(path=source)
-
-        index = clang.Index.create()
+def parse_source(source, compilation_database_path=None):
```
`parse_file(filename, compilation_database_path)` instead? It'll make this function in line with `get_compilation_commands`. (Doesn't matter which one you change, your own use should be consistent)

diivm replied:
Done.

diivm committed:
[parse.py] print_ast was at wrong pos

diivm committed:
[parse.py] use node dictionary

diivm committed:
[parse.py] change to parse_file

diivm committed:
[parse.py] More comments

diivm committed:
[utils.py] More comments

diivm committed:
add context

diivm committed:
[parse.py] fix import to work across directories

kunaltyagi commented on bindings/python/scripts/context.py:
```diff
@@ -0,0 +1,6 @@
+import os
+import sys
+
+sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))
```
Python's normal import not working for you?

diivm replied:
Nope. While running tests, it gives no `utils` for `parse.py`. If I replace it by `import scripts.utils` in `parse.py`, the tests work but the file itself doesn't (no `scripts`).

kunaltyagi replied:
Then you need to revist this later and reorganize the code. For now, it's ok, but please be on lookout for how to actually write module based scripts (like [venv](https://github.com/python/cpython/tree/master/Lib/venv): `python -m venv {args}`)

diivm replied:
I referred [this](https://docs.python-guide.org/writing/structure/) ([context topic](https://docs.python-guide.org/writing/structure/#test-suite))

It can also be solved by just adding the project to PYTHONPATH env, but I did this for now.

kunaltyagi commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,201 @@
     for child in cursor.get_children():
-        if node_in_this_file(child, this_filename):
-            walk_and_print(child, filter, lines, this_filename, depth + 1)
+        child_node = {"cursor": child, "filename": filename, "depth": depth + 1}
```
You're still operating under the assumption of `cursor` and `filename` being independent.

Why not:
```python
def print_ast(some_node):
    if some_node.cursor.spelling:
        print(some_node.depth, ...)
    for child in some_node.valid_children(): # or ...
    # for child in valid_children(some_node):
        print_ast(child)
```

diivm replied:
I misunderstood. I'll make the changes.

diivm replied:
Better structure. Added the `valid_children(node)` function.

kunaltyagi commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,201 @@
+        - depth: The depth of the node (root=0)
 
+    Returns:
+        - parsed_info (dict): 
```
Linters find it very difficult to help debug dicts. Since you know the data members already, use a class and if needed, make it interoperable with dict. That gives it way more power than current usage: https://stackoverflow.com/questions/35282222/

diivm replied:
#14 
New PR.

kunaltyagi commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,201 @@
+        if is_node_in_this_file(**child_node):
+            child_parsed_info = generate_parsed_info(**child_node)
+            # If both child and parent's info is not empty, add child's info to parent's members
+            if child_parsed_info and parsed_info:
```
* Can the `parsed_info` be None and `child_parsed_info` be non None?
* Since this is `and`, we can move check for `parsed_info` outside the loop
* Since `parsed_info` depends only on `cursor.spelling`, we can have an early exit, resulting in out-denting of lines 83-98 (incoming)

divmadan replied:
> Can the parsed_info be None and child_parsed_info be non None?

Yes for now. But that's an error since the parsed_info shouldn't be None #11. The `if cursor.spelling` shouldn't be checked. Needs #11 fixing.

> Can the parsed_info be None and child_parsed_info be non None?
> Since parsed_info depends only on cursor.spelling, we can have an early exit, resulting in out-denting of lines 83-98 (incoming)

Needs #11 fixing

divmadan replied:
Added:
```
    """
    - compilation_commands:
        - An iterable object providing all the compilation commands available to build filename.
        - type: <class 'clang.cindex.CompileCommands'>
    - compilation_commands[0]:
        - Since we have only one command per filename in the compile_commands.json, extract 0th element
        - type: <class 'clang.cindex.CompileCommand'>
    - compilation_commands[0].arguments:
        - Get compiler arguments from the CompileCommand object
        - type: <class 'generator'>
    - list(compilation_commands[0].arguments)[1:-1]:
        - Convert the generator object to list, and extract compiler arguments
        - 0th element is the compiler name
        - nth element is the filename
    """
```

diivm replied:
> Can the parsed_info be None and child_parsed_info be non None?

Yes for now, but that's an error that needs resolving #11. Will update after #11. (The is `cursor.spelling` was used to ignore empty symbols, but the unnamed structs, enums, etc. requires special treatment)

> Since this is and, we can move check for parsed_info outside the loop
> Since parsed_info depends only on cursor.spelling, we can have an early exit, resulting in out-denting of lines 83-98 (incoming)

Updation after #11

diivm replied:
#15 
New PR.

kunaltyagi commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,201 @@
+    compilation_commands = compilation_database.getCompileCommands(filename=filename)
 
+    # Extracting argument list from the command's generator object
+    compilation_commands = list(compilation_commands[0].arguments)[1:-1]
```
Let me ask about the reasoning behind 0, 1 and -1. You've added comments of **What** you're doing (eg: `Get compiler arguments from the compilation database for the given file`), but you're missing **Why** you're doing that thing in this way.

Eg:
* Why you didn't use `compile_commands, _ = db.getCompileCommands(name)`
* Why are the first and last commands useless?

diivm replied:
Added:
```
    """
    - compilation_commands:
        - An iterable object providing all the compilation commands available to build filename.
        - type: <class 'clang.cindex.CompileCommands'>
    - compilation_commands[0]:
        - Since we have only one command per filename in the compile_commands.json, extract 0th element
        - type: <class 'clang.cindex.CompileCommand'>
    - compilation_commands[0].arguments:
        - Get compiler arguments from the CompileCommand object
        - type: <class 'generator'>
    - list(compilation_commands[0].arguments)[1:-1]:
        - Convert the generator object to list, and extract compiler arguments
        - 0th element is the compiler name
        - nth element is the filename
    """
```

kunaltyagi commented on bindings/python/scripts/utils.py:
```diff
@@ -0,0 +1,88 @@
+    """
+
+    # pcl_path: contains the path as seen in the pcl directory
+    _, pcl_path = source.split(f"pcl{os.sep}", 1)
```
Make pcl an input argument

aPonza commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,201 @@
+    if cursor.location.file and cursor.location.file.name == filename:
+        return True
+    else:
+        return False
```
I could swear this used to be `return cursor.location.file and cursor.location.file.name == filename` and I wouldn't see a reason to change it

diivm replied:
Oh :joy: 
Changed this while writing the test that no longer exists. https://github.com/divmadan/pcl/pull/6#discussion_r450210107

Updated to:
```py
    if cursor.location.file is not None:
        return cursor.location.file.name == filename
    return False
```
for better intuition

aPonza commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,201 @@
+        - None
+    """
+
+    # Check if the cursor has spelling (cursor.spelling -> not NoneType)
     if cursor.spelling:
```
make the comment redundant by changing this to ` if cursor.spelling is not None:`, unless [you really mean to use that test](https://stackoverflow.com/questions/7816363/if-a-vs-if-a-is-not-none)

kunaltyagi replied:
Nice link :heart:

diivm replied:
spelling is string, changed to: `if cursor.spelling != ""`

aPonza commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,201 @@
+    parsed_info = dict()
+
+    # Check if the cursor has spelling (cursor.spelling -> not NoneType)
     if cursor.spelling:
```
make the comment redundant by changing this to ` if cursor.spelling is not None:`, unless [you really mean to use that test](https://stackoverflow.com/questions/7816363/if-a-vs-if-a-is-not-none)

aPonza commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,201 @@
+    # Dictionary containing parsed information of the source file
+    parsed_info = generate_parsed_info(**root_node)
+
+    return parsed_info
```
`return generate_parsed_info(**root_node)`

diivm committed:
[utils.py] modify to add generate.py functionalities

diivm committed:
[parse.py] update output_filepath

kunaltyagi commented on bindings/python/scripts/utils.py:
```diff
@@ -33,32 +33,44 @@ def get_json_output_path(source, output_dir):
     # filename: contains the output json's file name
     filename, _ = source_filename.split(".")
-    filename = f"{filename}.json"
+    filename = f"{filename}.{extension}"
```
Extension for json (as per docstring) is `.json`

diivm committed:
[parse.py] update function definitions and calls to use `node`

diivm committed:
[parse.py] updates and fixes

- add valid_children function
- update is_node_in_this_file, more intuitive
- improve comments

diivm committed:
[utils.py] extension updation and comments

diivm committed:
[parse.py] extension change

diivm committed:
[parse.py] return fix

kunaltyagi commented:
Tips for future:
* If PRs get too big, split them and merge the uncontroversial parts
* The introduction part of the PR should not have a laundry list of changes (see previous point). It makes it difficult to access what's good to merge, what's not (since it's possible to touch multiple lines in a PR in several commits)

Using git-flow helps in that regard:
* Base branch: master
* Split m-times at a commit to form devel (say devel-parser)
* Split at devel n-times to create a feature (say feature/docstring, feature/print_node)
* Create a PR from devel to master
* Create a PR from feature to devel

Given the response by @aPonza and myself, I'd say the lifetime of a feature branch should be at max 2 weeks (if possible, 3-4 days), and a devel branch should last a bit longer. Bring-up outstanding features to the notice in the discussion channel to find the solution to the issues worth discussing

aPonza commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,256 @@
-        print_node(cursor, lines, depth)
+    Arguments:
+        - compilation_database_path: The path to `compile_commands.json`
+        - filename: The file's name to get it's compilation commands
```
its

aPonza commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,256 @@
 
+def get_compilation_commands(compilation_database_path, filename):
+    """
+    Returns the compilation commands extracted from the the compile command
```
you probably meant "... from the compile_command.json"

aPonza commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,256 @@
+    for child_node in valid_children(node):
+        child_parsed_info = generate_parsed_info(child_node)
+        # If both child and parent's info is not empty (opening check for dictionary population), add child's info to parent's members
+        if child_parsed_info and parsed_info:
```
This is more pythonic than testing for `len(x) != 0`, well done

diivm replied:
Thank you!

aPonza commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,256 @@
-        holder["members"] = []
+    # Extracting argument list from the command's generator object
+    compilation_commands = list(compilation_commands[0].arguments)[1:-1]
 
-        parsed_list.append(holder)
+    return compilation_commands
```
`return list(compilation_commands[0].arguments)[1:-1]`

aPonza commented on bindings/python/scripts/parse.py:
```diff
@@ -1,145 +1,256 @@
-    filename = x_list[-1].split(".")[0]
-    relative_dir = "/".join(x for x in x_list[:-1] if x not in extra)
-    dir = os.path.join(output_dir, relative_dir)
+    # Is this check needed?
+    if compilation_database_path is None:
+        # Default compiler argument
+        compilation_commands = ["-std=c++14"]
```
This was for when tests didn't have a mock compile_commands.json iirc, so you could remove it here or leave it and remove it in #6.

aPonza commented on bindings/python/scripts/utils.py:
```diff
@@ -0,0 +1,114 @@
+
+    Arguments:
+        - source: The source's file name
+        - output_dir: The output_directory to write the json output
```
output directory (no underscore)

aPonza commented on bindings/python/scripts/utils.py:
```diff
@@ -0,0 +1,114 @@
+    # make the output directory if it doesn't exist
+    ensure_dir_exists(dir)
+
+    output_path = get_realpath(join_path(dir, filename))
+
+    return output_path
```
this one makes sense

aPonza reviewed (approved):
Honestly only nitpicks. Not so sure I like `parse_arguments` as is but the easy alternative is mostly the same so ok.

diivm committed:
[parse.py] small changes

diivm committed:
[utils.py] comment fix

divmadan commented:
@aPonza 
Thanks for the nitpicks :smiley: 
> Not so sure I like parse_arguments as is but the easy alternative is mostly the same so ok.

Do suggest. Will do that in another PR. 

Merging this for now.
